File: core-vm/src/native/java/lang/System.c
/**************************************************************************
* Parts copyright (c) 2001, 2002, 2003 by Punch Telematix.                *
* All rights reserved.                                                    *
* Parts copyright (c) 2004, 2005, 2006 by Chris Gray, /k/ Embedded Java   *
* Solutions. All rights reserved.                                         *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

#include <string.h>
#include <unistd.h>

#include "arrays.h"
#include "clazz.h"
#include "core-classes.h"
#include "descriptor.h"
#include "exception.h"
#include "environment.h"
#include "fastcall.h"
#include "heap.h"
#include "loading.h"
#include "checks.h"
#include "ts-mem.h"
#include "oswald.h"
#include "wstrings.h"
#include "threads.h"
#include "wonkatime.h"

w_long System_static_currentTimeMillis(JNIEnv *env, w_instance classSystem) {
  return (w_long)getNativeSystemTime();
}

void fast_System_static_currentTimeMillis(w_frame frame) {
  // Volatile 'coz gcc might screw up otherwise
  volatile union {w_long l; w_word w[2];} millis;

  millis.l = getNativeSystemTime();
  frame->jstack_top[0].s = 0;
  frame->jstack_top[0].c = millis.w[0];
  frame->jstack_top[1].s = 0;
  frame->jstack_top[1].c = millis.w[1];
  frame->jstack_top += 2;
  woempa(1, "returning 0x%08x 0x%08x\n", frame->jstack_top[-2].c, frame->jstack_top[-1].c);
}

static void do_boolean_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_byte * src;
  w_int    srcMask;
  w_byte * dst;
  w_int    dstMask;
  w_int    i;

  if (srco > dsto) {
    src = ((w_byte *)(Src + F_Array_data)) + srco / 8;
    dst = ((w_byte *)(Dst + F_Array_data)) + dsto / 8;
    srcMask = 1 << (srco % 8);
    dstMask = 1 << (dsto % 8);

    for (i = 0; i < length; i++) {
      if (*src & srcMask) {
        *dst |= dstMask;
      }
      else {
        *dst &= 0xffffffff ^ dstMask;
      }

      srcMask <<= 1;
      if (srcMask == 0x100) {
        src++;
        srcMask = 1;
      }

      dstMask <<= 1;
      if (dstMask == 0x100) {
        dst++;
        dstMask = 1;
      }
    }
  }
  else {
    length -= 1;
    src = ((w_byte *)(Src + F_Array_data)) + (srco + length) / 8;
    dst = ((w_byte *)(Dst + F_Array_data)) + (dsto + length) / 8;
    srcMask = 1 << ((srco + length) % 8);
    dstMask = 1 << ((dsto + length) % 8);

    for (i = 0; i <= length; i++) {
      if (*src & srcMask) {
        *dst |= dstMask;
      }
      else {
        *dst &= 0xffffffff ^ dstMask;
      }

      srcMask >>= 1;
      if (srcMask == 0 ) {
        src--;
        srcMask = 0x80;
      }

      dstMask >>= 1;
      if (dstMask == 0) {
        dst--;
        dstMask = 0x80;
      }
    }
  }

}

static inline void do_byte_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_ubyte * src = ((w_byte *)(Src + F_Array_data)) + srco;
  w_ubyte * dst = ((w_byte *)(Dst + F_Array_data)) + dsto;
  /* [CG 20050620] Dunno about all this Duff stuff, what say we just do a memcpy or a memmove?
  w_int duffs = (length + 15) / 16;

  if (src < dst && dst < src + length) {
    dst += length;
    src += length;
    switch (length & 0x0f) {
      case  0: do { *--dst = *--src;
      case 15:      *--dst = *--src;
      case 14:      *--dst = *--src;
      case 13:      *--dst = *--src;
      case 12:      *--dst = *--src;
      case 11:      *--dst = *--src;
      case 10:      *--dst = *--src;
      case  9:      *--dst = *--src;
      case  8:      *--dst = *--src;
      case  7:      *--dst = *--src;
      case  6:      *--dst = *--src;
      case  5:      *--dst = *--src;
      case  4:      *--dst = *--src;
      case  3:      *--dst = *--src;
      case  2:      *--dst = *--src;
      case  1:      *--dst = *--src;
              } while (--duffs > 0);
    }
  }
  else {
    switch (length & 0x0f) {
      case  0: do { *dst++ = *src++;
      case 15:      *dst++ = *src++;
      case 14:      *dst++ = *src++;
      case 13:      *dst++ = *src++;
      case 12:      *dst++ = *src++;
      case 11:      *dst++ = *src++;
      case 10:      *dst++ = *src++;
      case  9:      *dst++ = *src++;
      case  8:      *dst++ = *src++;
      case  7:      *dst++ = *src++;
      case  6:      *dst++ = *src++;
      case  5:      *dst++ = *src++;
      case  4:      *dst++ = *src++;
      case  3:      *dst++ = *src++;
      case  2:      *dst++ = *src++;
      case  1:      *dst++ = *src++;
              } while (--duffs > 0);
    }
  }
  */
  if (Src == Dst) {
    memmove(dst, src, length);
  }
  else {
    w_memcpy(dst, src, length);
  }
}

static void do_16bits_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_short * src = ((w_short *)(Src + F_Array_data)) + srco;
  w_short * dst = ((w_short *)(Dst + F_Array_data)) + dsto;
  w_int duffs = (length + 15) / 16;

  if (src < dst && dst < src + length) {
    dst += length;
    src += length;
    switch (length & 0x0f) {
      case  0: do { *--dst = *--src;
      case 15:      *--dst = *--src;
      case 14:      *--dst = *--src;
      case 13:      *--dst = *--src;
      case 12:      *--dst = *--src;
      case 11:      *--dst = *--src;
      case 10:      *--dst = *--src;
      case  9:      *--dst = *--src;
      case  8:      *--dst = *--src;
      case  7:      *--dst = *--src;
      case  6:      *--dst = *--src;
      case  5:      *--dst = *--src;
      case  4:      *--dst = *--src;
      case  3:      *--dst = *--src;
      case  2:      *--dst = *--src;
      case  1:      *--dst = *--src;
              } while (--duffs > 0);
    }
  }
  else {
    switch (length & 0x0f) {
      case  0: do { *dst++ = *src++;
      case 15:      *dst++ = *src++;
      case 14:      *dst++ = *src++;
      case 13:      *dst++ = *src++;
      case 12:      *dst++ = *src++;
      case 11:      *dst++ = *src++;
      case 10:      *dst++ = *src++;
      case  9:      *dst++ = *src++;
      case  8:      *dst++ = *src++;
      case  7:      *dst++ = *src++;
      case  6:      *dst++ = *src++;
      case  5:      *dst++ = *src++;
      case  4:      *dst++ = *src++;
      case  3:      *dst++ = *src++;
      case  2:      *dst++ = *src++;
      case  1:      *dst++ = *src++;
              } while (--duffs > 0);
    }
  }

}

static void do_32bits_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_word * src = ((w_word *)(Src + F_Array_data)) + srco;
  w_word * dst = ((w_word *)(Dst + F_Array_data)) + dsto;
  w_int duffs = (length + 15) / 16;

  if (src < dst && dst < src + length) {
    dst += length;
    src += length;
    switch (length & 0x0f) {
      case  0: do { *--dst = *--src;
      case 15:      *--dst = *--src;
      case 14:      *--dst = *--src;
      case 13:      *--dst = *--src;
      case 12:      *--dst = *--src;
      case 11:      *--dst = *--src;
      case 10:      *--dst = *--src;
      case  9:      *--dst = *--src;
      case  8:      *--dst = *--src;
      case  7:      *--dst = *--src;
      case  6:      *--dst = *--src;
      case  5:      *--dst = *--src;
      case  4:      *--dst = *--src;
      case  3:      *--dst = *--src;
      case  2:      *--dst = *--src;
      case  1:      *--dst = *--src;
              } while (--duffs > 0);
    }
  }
  else {
    switch (length & 0x0f) {
      case  0: do { *dst++ = *src++;
      case 15:      *dst++ = *src++;
      case 14:      *dst++ = *src++;
      case 13:      *dst++ = *src++;
      case 12:      *dst++ = *src++;
      case 11:      *dst++ = *src++;
      case 10:      *dst++ = *src++;
      case  9:      *dst++ = *src++;
      case  8:      *dst++ = *src++;
      case  7:      *dst++ = *src++;
      case  6:      *dst++ = *src++;
      case  5:      *dst++ = *src++;
      case  4:      *dst++ = *src++;
      case  3:      *dst++ = *src++;
      case  2:      *dst++ = *src++;
      case  1:      *dst++ = *src++;
              } while (--duffs > 0);
    }
  }

}

static void do_64bits_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {
  do_32bits_arraycopy(Src, srco, Dst, dsto, length * 2);
}

/*
** A checked assignment of references; return true if there was a problem...
*/

static inline w_boolean do_checked_reference_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_instance * src = ((w_instance *)(Src + F_Array_data)) + srco;
  w_instance * dst = ((w_instance *)(Dst + F_Array_data)) + dsto;
  w_clazz      dstComponent = instance2clazz(Dst)->previousDimension;
  w_int        i;

  if (src < dst && dst < src + length) {
    for (i = length - 1; i >= 0; i--) {
      if (src[i] == NULL || isAssignmentCompatible(instance2clazz(src[i]), dstComponent)) {
        setArrayReferenceField(Dst, src[i], dsto + i);
      }
      else {
        return TRUE;
      }
    }
  }
  else {
    for (i = 0; i < length; i++) {
      if (src[i] == NULL || isAssignmentCompatible(instance2clazz(src[i]), dstComponent)) {
        setArrayReferenceField(Dst, src[i], dsto + i);
      }
      else {
        return TRUE;
      }
    }
  }

  return FALSE;

}

/*
 * [ORIGINAL COMMENT]
** In the reference copying, we can safely ignore setArrayReferenceField since the references
** are protected allready by the source array.
** [CG 20050620] Not so, my friend. This code is also used when the source array
** and the destination array are one and the same, and in that case we have a
** real risk of a race condition (imagine that the array is being marked by
** another thread while we are busy shuffling data,
** Therefore I have switched to using a memcpy/memmove, wrapped in an "unsafe" 
** context.
*/

static inline void do_unchecked_reference_arraycopy(w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_instance * src = ((w_instance *)(Src + F_Array_data)) + srco;
  w_instance * dst = ((w_instance *)(Dst + F_Array_data)) + dsto;
  /* WAS:
  w_int duffs = (length + 15) / 16;

  if (src < dst && dst < src + length) {
    dst += length;
    src += length;
    switch (length & 0x0f) {
      case  0: do { *--dst = *--src;
      case 15:      *--dst = *--src;
      case 14:      *--dst = *--src;
      case 13:      *--dst = *--src;
      case 12:      *--dst = *--src;
      case 11:      *--dst = *--src;
      case 10:      *--dst = *--src;
      case  9:      *--dst = *--src;
      case  8:      *--dst = *--src;
      case  7:      *--dst = *--src;
      case  6:      *--dst = *--src;
      case  5:      *--dst = *--src;
      case  4:      *--dst = *--src;
      case  3:      *--dst = *--src;
      case  2:      *--dst = *--src;
      case  1:      *--dst = *--src;
              } while (--duffs > 0);
    }
  }
  else {
    switch (length & 0x0f) {
      case  0: do { *dst++ = *src++;
      case 15:      *dst++ = *src++;
      case 14:      *dst++ = *src++;
      case 13:      *dst++ = *src++;
      case 12:      *dst++ = *src++;
      case 11:      *dst++ = *src++;
      case 10:      *dst++ = *src++;
      case  9:      *dst++ = *src++;
      case  8:      *dst++ = *src++;
      case  7:      *dst++ = *src++;
      case  6:      *dst++ = *src++;
      case  5:      *dst++ = *src++;
      case  4:      *dst++ = *src++;
      case  3:      *dst++ = *src++;
      case  2:      *dst++ = *src++;
      case  1:      *dst++ = *src++;
              } while (--duffs > 0);
    }
  }
  */

  w_thread thread = currentWonkaThread;
  w_boolean unsafe = enterUnsafeRegion(thread);
  if (Src == Dst) {
    memmove(dst, src, length * sizeof(w_instance));
  }
  else {
    memcpy(dst, src, length * sizeof(w_instance));
  }
  if (!unsafe) {
    enterSafeRegion(thread);
  }
}

void System_static_arraycopy(JNIEnv *env, w_instance Class, w_instance Src, w_int srco, w_instance Dst, w_int dsto, w_int length) {

  w_thread thread = JNIEnv2w_thread(env);
  w_clazz   srcClazz;
  w_clazz   dstClazz;
  w_clazz   srcCom;
  w_clazz   dstCom;
  w_int     srcLength;
  w_int     dstLength;
  w_boolean problem = FALSE;

  /*
  ** Do the required checks...
  */
  
  if (Src == NULL || Dst == NULL) {
    throwException(thread, clazzNullPointerException, NULL);
    return;
  }

  srcClazz = instance2clazz(Src);
  dstClazz = instance2clazz(Dst);

  if (!srcClazz->dims || !dstClazz->dims) {
    throwException(thread, clazzArrayStoreException, NULL);
    return;
  }

  srcCom = srcClazz->previousDimension;
  dstCom = dstClazz->previousDimension;

  if (clazzIsPrimitive(srcCom) && clazzIsPrimitive(dstCom) && srcCom != dstCom) {
    problem = TRUE;
  }
  else if (clazzIsPrimitive(srcCom) && !clazzIsPrimitive(dstCom)) {
    problem = TRUE;
  }
  else if (!clazzIsPrimitive(srcCom) && clazzIsPrimitive(dstCom)) {
    problem = TRUE;
  }

  if (problem) {
    throwException(thread, clazzArrayStoreException, NULL);
    return;
  }

  srcLength = instance2Array_length(Src);
  dstLength = instance2Array_length(Dst);

  if (srco < 0 || dsto < 0 || length < 0 || srco > srcLength - length || dsto > dstLength - length) {
    throwException(thread, clazzArrayIndexOutOfBoundsException, NULL);
    return;
  }

  /*
  ** Eliminate trivial cases.
  */

  if (length == 0 || ((Src == Dst) && (srco == dsto))) {
    return;
  }

  /*
  ** Call the appropriate routine, depending on destination and source type. If any assignment
  ** conversion is required, it is done in special routines.
  */

  if (clazzIsPrimitive(srcCom)) {
    switch (srcCom->type & 0x0f) {
      case VM_TYPE_BOOLEAN: {
        do_boolean_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_CHAR: {
        do_16bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }

      case VM_TYPE_SHORT: {
        do_16bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_FLOAT: {
        do_32bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_INT: {
        do_32bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_DOUBLE: {
        do_64bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_LONG: {
        do_64bits_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      case VM_TYPE_BYTE: {
        do_byte_arraycopy(Src, srco, Dst, dsto, length);
        break;
      }
      
      default: break;
      
    }
  }
  else {
    if (dstClazz == srcClazz || dstClazz == clazzArrayOf_Object) {
      do_unchecked_reference_arraycopy(Src, srco, Dst, dsto, length);
    }
    else {
      if (do_checked_reference_arraycopy(Src, srco, Dst, dsto, length)) {
        throwException(thread, clazzArrayStoreException, NULL);
      }
    }
  }

}

w_int System_identityHashCode(JNIEnv *env, jclass clazz, w_instance obj) {

  if (obj == NULL) {
    return 0;
  }

  return Object_hashCode(env, obj);

}

extern Wonka_InitArgs *system_vm_args;

w_instance System_getCmdLineProperties(JNIEnv *env, w_instance this) {

  w_thread    thread = JNIEnv2w_thread(env);
  w_int       i;
  w_instance  Array = NULL;
  w_instance  String;
  w_string    string;
  w_int       length;

  for (length = 0; length < 100; length++) {
    if (system_vm_args->properties[length] == NULL) {
      break;
    }
  }

  if (length > 0) {

    Array = allocArrayInstance_1d(thread, clazzArrayOf_String, length);

    if (Array) {
      for (i = 0; i < length; i++) {
        String = allocInstance_initialized(thread, clazzString);
        if (String) {
          string = cstring2String(system_vm_args->properties[i], strlen(system_vm_args->properties[i]));
          setWotsitField(String, F_String_wotsit, string);
        } 
        setArrayReferenceField(Array, String, i);
      }
    }
  }

  return Array;

}

File: Mika/trunk/core-vm/java/java/lang/System.java
/**************************************************************************
* Parts copyright (c) 2001, 2002, 2003 by Punch Telematix. All rights     *
* reserved.                                                               *
* Parts copyright (c) 2004, 2005, 2007, 2008 by Chris Gray, /k/ Embedded  *
* Java Solutions. All rights reserved.                                    *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

package java.lang;

import java.io.IOException;
import java.io.InputStream;
import java.io.PrintStream;
import java.lang.reflect.Field;
import java.util.Properties;
import java.util.PropertyPermission;

public final class System {

  public final static InputStream in;
  public final static PrintStream out;
  public final static PrintStream err;

  private static SecurityManager theSecurityManager;
  private static Runtime theRuntime;
  private static Properties systemProperties;
  private static boolean initialized;

  static {
    in = new wonka.io.StandardInputStream();
    // We make 'out' non-autoflush and 'err' autoflush, by analogy with Posix.
    out = new PrintStream(new wonka.io.StandardOutputStream(), false);
    err = new PrintStream(new wonka.io.ErrorOutputStream(), true);
    theRuntime = Runtime.getRuntime();
    systemProperties = new Properties(new DefaultProperties());

    try {
      InputStream syspropstream = ClassLoader.getSystemResourceAsStream("system.properties");
      if (syspropstream != null) {
        systemProperties.load(syspropstream);
      }
      String mikaprops = systemProperties.getProperty("mika.properties");
      if (mikaprops == null) {
        mikaprops = "mika.properties";
      }
      syspropstream = ClassLoader.getSystemResourceAsStream(mikaprops);
      if (syspropstream != null) {
        systemProperties.load(syspropstream);
      }
      // if not found we still set the 'loaded' property, as if we found
      // a trivial or empty file.
      systemProperties.setProperty("mika.properties.loaded", "true");
    }
    catch (IOException e) {}

    parseCmdLineProperties();

    try {    
      //FORCE SOME CALLS things that are postponed until the system properties were set up correctly
      ClassLoader.get_defaultProtectionDomain();
      ClassLoader.createApplicationClassLoader();

      //SETUP DEFAULT LOCALE
      java.util.Locale.setDefault(new java.util.Locale(getProperty("user.language", "en"), getProperty("user.country", "")));
      
      String theManager = getProperty("java.security.manager");
      if("".equals(theManager) || "default".equals(theManager)) {
        theSecurityManager = new SecurityManager();
      }
      else {
        // TODO: Use the given class..
      }
      
      initialized = true;
    }
    catch(Throwable t){
      t.printStackTrace();
    }

    if (!initialized) {
      exit(1);
    }
  }

  private static native String[] getCmdLineProperties();

  private static void parseCmdLineProperties() {
    String[] props = getCmdLineProperties();
    if(props != null) {
      for(int i=0; i < props.length; i++) {
        String key, value;
        int equals_at = props[i].indexOf('=');
        if (equals_at < 0) {
          key = props[i];
          wonka.vm.Etc.woempa(10,"SystemInit: Setting system property '"+key+"' to \"\"");
          systemProperties.put(key,"");
        }
        else if (equals_at == 0) {
          wonka.vm.Etc.woempa(10,"SystemInit: Ignoring nameless property in "+props[i]);
        }
        else if (equals_at == props[i].length()-1) {
          key = props[i].substring(0,equals_at);
          wonka.vm.Etc.woempa(10,"SystemInit: Setting system property '"+key+"' to \"\"");
          systemProperties.put(key,"");
        }
        else {
          key = props[i].substring(0, equals_at);
          value = props[i].substring(equals_at + 1);
          wonka.vm.Etc.woempa(10,"SystemInit: Setting system property '"+key+"' to '"+value+"'");
          systemProperties.put(key,value);
        }
      }
    }
  }

  public static String getenv(String name) {
    return getProperty(name);
  }
  
  public static void runFinalizersOnExit(boolean run) {
    Runtime.runFinalizersOnExit(run);
  }
  
  public static void setIn(InputStream newIn) {
    if (in!=null) {
      try {
        in.close();
      } catch (java.io.IOException e) {}
    }
    setField("in", newIn);
  }

  private static void setField(String name, Object value) {
    try {      
      Field field = System.class.getDeclaredField(name);
      field.setAccessible(true);
      try {
        field.set(null, value);
      }catch (IllegalAccessException e) {
        e.printStackTrace();
      }
    } catch (NoSuchFieldException e) {
      e.printStackTrace();
    }    
    
  }
  
  public static void setOut(PrintStream newOut) {
    if (out!=null) {
        out.close();
    }
    setField("out", newOut);
  }

  public static void setErr(PrintStream newErr) {
    if (err!=null) {
        err.close();
    }
    setField("err", newErr);
  }

  public static SecurityManager getSecurityManager() { 
    return theSecurityManager;
  }

  public static void setSecurityManager(SecurityManager sm)
    throws SecurityException
  { 
    if (theSecurityManager!=null) {
      if (wonka.vm.SecurityConfiguration.USE_ACCESS_CONTROLLER
       || wonka.vm.SecurityConfiguration.USE_SECURITY_MANAGER) {
        java.security.AccessController.checkPermission(new RuntimePermission("setSecurityManager"));
      }
    }
    theSecurityManager = sm;
  }

  public static native long currentTimeMillis();

  private static void propertyCheck(String propname) {
    if (wonka.vm.SecurityConfiguration.USE_ACCESS_CONTROLLER) {
      java.security.AccessController.checkPermission(new PropertyPermission(propname, "read"));
    }
    else if (wonka.vm.SecurityConfiguration.USE_SECURITY_MANAGER) {
      if (theSecurityManager != null) {
        theSecurityManager.checkPropertyAccess(propname);
      }
    }
  }

  private static void propertiesCheck() {
    if (wonka.vm.SecurityConfiguration.USE_ACCESS_CONTROLLER) {
      java.security.AccessController.checkPermission(new PropertyPermission("*", "read,write"));
    }
    else if (wonka.vm.SecurityConfiguration.USE_SECURITY_MANAGER) {
      if (theSecurityManager != null) {
        theSecurityManager.checkPropertiesAccess();
      }
    }
  }

  public static Properties getProperties() 
    throws SecurityException
  {
    propertiesCheck();

    return systemProperties;
  }

  public static void setProperties(Properties props) 
    throws SecurityException
  {
    propertiesCheck();

    systemProperties = props;
  }


  public static String getProperty(String key) 
    throws SecurityException
  {
    propertyCheck(key);
    if(key.equals("")) {
      throw new IllegalArgumentException();
    }
    return systemProperties.getProperty(key);
  }

    
  public static String getProperty(String key, String defaults)
    throws SecurityException
  {
    propertyCheck(key);

    try {

      return systemProperties.getProperty(key,defaults);

    }
    catch (NullPointerException npe) {

      return defaults;

    }
  }

  public static String setProperty(String key, String defaults)
    throws SecurityException
  {
    if (wonka.vm.SecurityConfiguration.USE_ACCESS_CONTROLLER) {
      java.security.AccessController.checkPermission(new PropertyPermission(key, "write"));
    }
    else if (wonka.vm.SecurityConfiguration.USE_SECURITY_MANAGER) {
      if (theSecurityManager != null) {
        theSecurityManager.checkPermission(new PropertyPermission(key, "write"));
      }
    }
    return (String)systemProperties.setProperty(key,defaults);
  }


  public static void exit(int status) 
    throws SecurityException
  {
    theRuntime.exit(status);
  }

  public static void gc() {
    theRuntime.gc();
  }

  public static void runFinalization() {
    theRuntime.runFinalization();
  }

  public static void load(String libname) 
    throws SecurityException, UnsatisfiedLinkError
  {
    throw new UnsatisfiedLinkError("not implemented");
  }

  public static void loadLibrary(String libname) 
    throws SecurityException, UnsatisfiedLinkError
  {
    // throw new UnsatisfiedLinkError("not implemented");
    Runtime.getRuntime().loadLibrary(libname);
  }
  
  public static String mapLibraryName(String libname) {
    return "lib" + libname + ".so";
  }

  native public static void arraycopy(Object src, int srcOffset, Object dst, int dstOffset, int length)
    throws NullPointerException, ArrayStoreException, ArrayIndexOutOfBoundsException;

  public static native int identityHashCode(Object o);

  private System() {}
}
